File: src/backend/app/schemas/skill.py
"""Pydantic schemas for the skill tree progression system."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel


class SkillNodeRead(BaseModel):
    id: str
    tree_id: str
    name: str
    description: str
    parent_id: Optional[str] = None
    level_required: int
    cost: int
    stat_bonus: Optional[Any] = None
    icon: str
    tier: int

    model_config = {"from_attributes": True}


class SkillTreeRead(BaseModel):
    id: str
    name: str
    description: str
    category: str
    icon: str
    nodes: list[SkillNodeRead] = []

    model_config = {"from_attributes": True}


class AgentSkillRead(BaseModel):
    id: str
    agent_id: str
    node_id: str
    unlocked_at: datetime
    node: Optional[SkillNodeRead] = None

    model_config = {"from_attributes": True}


class UnlockSkillResponse(BaseModel):
    agent_skill_id: str
    node_id: str
    tokens_spent: int
    new_wallet_balance: int

File: src/frontend/src/types/skill.ts
export interface SkillNode {
  id: string;
  tree_id: string;
  name: string;
  description: string;
  parent_id: string | null;
  level_required: number;
  cost: number;
  stat_bonus: Record<string, number>;
  icon: string;
  tier: number;
}

export interface SkillTree {
  id: string;
  name: string;
  description: string;
  category: string;
  icon: string;
  nodes: SkillNode[];
}

export interface AgentSkill {
  id: string;
  agent_id: string;
  node_id: string;
  unlocked_at: string;
}

export type SkillNodeState = 'locked' | 'available' | 'unlocked';
